### PR TITLE
Ndebug refactoring

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This uses the newly-extracted https://github.com/atlas-engineer/ndebug for custom debugging purposes of Nyxt. Not much there, mostly some small optimizations and saved lines of code :)

The dependencies of `ndebug` and, consequently, Nyxt, now include `trivial-custom-debugger`.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.  

- [X] I have pulled from master before submitting this PR
- [X] My code follows the style guidelines for Common Lisp code
  - See [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf) and [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml).
- [X] I have performed a self-review of my own code
- [x] My code has been reviewed by at least one peer (the peer review to approve a PR counts.  The reviewer must download and test the code)
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
  -  Will expand the manual on that.
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
  - Not sure about those. How do we test the debugger? 
- [X] New and existing unit tests pass locally with my changes
- [X] There are no merge conflicts
- [x] Send a patch to Guix.

### CODE REVIEW

**Comment by aadcg** [comment]
Nice!

As far as I can understand, the selected region of the backtrace below is not very useful. Would it be possible to remove it?

**Comment by aadcg** [comment]
I've added some minor niceties to the interface.

**Comment by aadcg** [comment]
Regarding testing, I think we should add basic tests to `ndebug`. I don't see any point in testing from Nyxt's side.

File: source/debugger.lisp [hunk 1]
```diff
@@ -61,20 +61,20 @@ See `ndebug:condition-wrapper' for documentation."))
...
-                         when (or (typep arg 'dissect:unknown-arguments)
-                                  (typep arg 'dissect:unavailable-argument))
-                           collect (:li (:code "Unknown argument"))
-                         else collect (:li (:raw (value->html arg t)))))))))))))
+    (dolist (frame (ndebug:stack wrapper))
+      (let ((frame-call (dissect:call frame))
+            (frame-args (dissect:args frame)))
+        (when frame-call
+          (:details
+           :attrs (list :open frame-args)
```
[the remark this answers is not in the document]
Comment (aartaka) [reply source/debugger.lisp:69]: As an alternative, don't make it a `:details` when there's no arguments. That's what I wanted to do just an hour ago :P

File: source/mode/repl.lisp [hunk 2]
```diff
@@ -137,6 +137,8 @@ Features:
              :padding "3px"
              :autofocus "true"
              :width "100%")
+            (.button
+             :display "block")
```
Comment (aartaka) [inline source/mode/repl.lisp:141]: This part of our style (reflowing buttons) is quite useful in Gopher and other interfaces in Nyxt, as it spreads out nicely across the allocated space. Making buttons block-only takes too much space for what buttons pursue here -- quick immediate and mutually exclusive action.

There are benefits to the block view (like being more predictable and structured), but it's more consistent and appealing (to me) to cluster things :P

**Comment by aartaka** [comment]
> Regarding testing, I think we should add basic tests to `ndebug`. I don't see any point in testing from Nyxt's side.

Yes, that I agree with.

**Comment by aartaka** [comment]
> As far as I can understand, the selected region of the backtrace below is not very useful. Would it be possible to remove it?

While it's not exactly useful, that's what stack looks like when we request all of it in the debugger hook.

I can agree with removing just the first four frames (as they are directly related to the debugger), but no more/ The frames after the fourth can have some useful information too.

**Comment by aartaka** [comment]
I've pushed some opinionated changes too :P

**Comment by aartaka** [comment]
@aadcg, looks bearable to you?

Would you be willing to send patch to Guix, or should I finally step out of my comfort zone and try doing it myself?

File: source/debugger.lisp [hunk 3]
```diff
@@ -61,20 +61,20 @@ See `ndebug:condition-wrapper' for documentation."))
...
-              (:p "Called with:")
-              (:ul (loop for arg in (dissect:args frame)
-                         when (or (typep arg 'dissect:unknown-arguments)
-                                  (typep arg 'dissect:unavailable-argument))
-                           collect (:li (:code "Unknown argument"))
-                         else collect (:li (:raw (value->html arg t)))))))))))))
+    (dolist (frame (ndebug:stack wrapper))
+      (let ((frame-call (dissect:call frame))
+            (frame-args (dissect:args frame)))
+        (when frame-call
```
[the remark this answers is not in the document]
Comment (aadcg) [reply source/debugger.lisp]: Does such a case actually exist?

File: source/debugger.lisp (hunk 1 above)
[the remark this answers is not in the document]
Comment (aadcg) [reply source/debugger.lisp]: > REPL debugger is there mostly for fast fixes and immediately useful information. Spreading it out makes it too long of an interface to be useful from the fist glance :(

Highly debatable actually. Think of the kind of backtraces that Python throws for instance. But yes, there should be a configurable slot.

[the remark this answers is not in the document]
Comment (aadcg) [reply source/debugger.lisp]: > As an alternative, don't make it a `:details` when there's no arguments. That's what I wanted to do just an hour ago :P

Of course... Silly me.

**Comment by aadcg** [comment]
> Would you be willing to send patch to Guix, or should I finally step out of my comfort zone and try doing it myself?

As you wish really.

There's nothing special to it actually! I can share with you how I usually do it :)

File: source/mode/repl.lisp [hunk 4]
```diff
@@ -318,13 +318,15 @@ Features:
             (raised-condition evaluation))
        (let ((wrapper (raised-condition evaluation)))
          (:pre (format nil "~a" (ndebug:condition-itself wrapper)))
+         (:h3 "Restarts:")
```
[the remark this answers is not in the document]
Comment (aadcg) [reply source/mode/repl.lisp]: Fine for me.

File: source/mode/repl.lisp (hunk 2 above)
Comment (aadcg) [reply source/mode/repl.lisp:141]: I can't agree that reflowing buttons is a good default in this case.

But I doesn't bother me too much.

**Comment by aadcg** [comment]
Everything looks good to me but I'd like to add a user option to automatically expand the backtrace and sort out [this question](https://github.com/atlas-engineer/nyxt/pull/2461#discussion_r922829992).

File: source/debugger.lisp (hunk 3 above)
[the remark this answers is not in the document]
Comment (aartaka) [reply source/debugger.lisp]: I don't know, but better account for it -- compilers are quire probable to create unnamed function blocks, for one :)

**Comment by aartaka** [comment]
OK, NDebug is mostly ready and will likely stay the way it is now (API-wise) for long enough to depend on it. Nyxt-side of the debugger works fine too. I'll add some tests to NDebug and write a short manual section regarding debugger.

In the meanwhile, @aadcg, any other changes you'd make (both on the NDebug and Nyxt debugger sides)? Auto-expanding the view as a configurable option sounds reasonable enough, for one :)

**Comment by Ambrevar** [comment]
I've added trivial-custom-debugger to Guix.

**Comment by aartaka** [comment]
OK, I've added a small manual section regarding the debugger and newly-added `report-bug` command.

**Comment by aartaka** [comment]
@Ambrevar, ndebug is Guix-ready, I suppose! It has tests and documentation, and it the API seems to be usable. Would you mind sending a patch to Guix? :)

**Comment by Ambrevar** [comment]
Will do as soon as possible!

**Comment by Ambrevar** [comment]
@aartaka I'm ready to commit the Ndebug package to Guix.  Should I do it on 0.1.0 or you want to push something else?

**Comment by aartaka** [comment]
> @aartaka I'm ready to commit the Ndebug package to Guix. Should I do it on 0.1.0 or you want to push something else?

I believe it's ready. The Swank dependency is too complex of an issue to rush with it, so let's leave it for later.

**Comment by Ambrevar** [comment]
Done!

**Comment by aartaka** [comment]
Merged with 190ea36c9ab7795088fa3287c41fb418d80ef32b!

**Pull request merged**